### PR TITLE
azurerm_monitor_private_link_scope - api 2021-07-01-preview - ingestion and query mode access added

### LINKED ISSUE #10059: Support for Azure Monitor Private Link Scope
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Allow azurerm provider to create/update Azure Monitor Private Link Scope resources: [az cli reference](https://docs.microsoft.com/en-us/cli/azure/monitor/private-link-scope?view=azure-cli-latest). This is supported in the 2019-10-17 preview schema [here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/monitor/resource-manager/Microsoft.Insights/preview/2019-10-17-preview/privateLinkScopes_API.json). But I don't believe it is in the Azure Go SDK.
 
Implementing this will enable use of Azure Monitor behind Private Endpoints.

### New or Affected Resource(s)

* azurerm_monitor

### Potential Terraform Configuration

```hcl
resource "azurerm_monitor_private_link_scope" "example" {
  name = "example"
  log_analytics_workspace_id = "law_id"
}
```

### LINKED ISSUE #19370: Attribute missing from azure monitor private link scope resource
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

0.14.10

### AzureRM Provider Version

2.98.0

### Affected Resource(s)/Data Source(s)

azurerm_monitor_private_link_scope

### Terraform Configuration Files

```hcl
## source module
resource "azurerm_monitor_private_link_scope" "monitor_private_link_scope" {
  name                = join("-", ["ampls", var.project, var.workload, var.environment, var.location, var.padding]) 
  resource_group_name = var.resource_group_name
  tags                = var.default_tags
}

# Create Azure Monitor Private Link Scope resources
module "monitor-private-link-scope" {
  source              = ".....azurerm/Azure-Monitor-Private-Link-Scope"
  project             = var.project
  environment         = var.environment
  location            = var.location
  workload            = var.private_link_scope_workload_name
  padding             = var.padding
  resource_group_name = module.resource-group.resource_group_name
  default_tags        = local.default_tags
  depends_on = [
    module.resource-group
  ]
}
```

### Debug Output/Panic Output

```shell
-
```

### Expected Behaviour

azurerm_monitor_private_link_scope should provide the user to select the AMPLS access modes. Access modes `Ingestion access mode` and `Query access mode` user should be able to set these values from TF module, but the required attributes were not provided. in the module https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_private_link_scope it only provides the AMPLS name and resource group only. Users should be able to changes these values via terraform.

### Actual Behaviour

in the `azurerm_monitor_private_link_scope` TF module AMPLS access mode configuration was not provided and it will initially set `Ingestion access mode` to private only and `Query access mode` to open, 

### Steps to Reproduce

1. create the TF module.
```
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_monitor_private_link_scope" "example" {
  name                = "example-ampls"
  resource_group_name = azurerm_resource_group.example.name
}
```
2. Run the terraform apply command.

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by KiPIDesTAN**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## Description

Updated Azure Monitor Private Link Scope to the latest API, 2021-07-01-preview, to support ingestion mode access and query mode access. Fixes hashicorp/terraform-provider-azurerm#19370 and fixes hashicorp/terraform-provider-azurerm#10059.

This item adds ingestion_access_mode and query_access_mode items as required items to the azurerm_monitor_private_link_scope resource. 

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [X] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [X] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [X] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.

## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

* `azurerm_monitor_private_link_scope` - support for the `ingestion_access_mode` and `query_access_mode` property [GH-19370]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #19370
Fixes #10059

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/monitor/monitor_private_link_scope_resource.go [hunk 1]
```diff
@@ -47,6 +48,28 @@ func resourceMonitorPrivateLinkScope() *pluginsdk.Resource {
...
+						string(privatelinkscopesapis.AccessModeOpen),
+						string(privatelinkscopesapis.AccessModePrivateOnly),
+					},
+					false),
+			},
+
+			"query_access_mode": {
+				Type:     pluginsdk.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice(
```
Comment (katbyte) [inline internal/services/monitor/monitor_private_link_scope_resource.go:65]: adding these is a rbeaking change? shouldn't they be optional?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @KiPIDesTAN - overall looks good but i think this is a breaking change for users with these properties being optiona;?

File: internal/services/monitor/monitor_private_link_scope_resource.go (hunk 1 above)
Comment (KiPIDesTAN) [reply internal/services/monitor/monitor_private_link_scope_resource.go]: @katbyte , good catch. 

My most recent push on this PR has changes to make both ingestion_access_mode and query_access_mode optional with default values of 'Open'. Would you mine reviewing again, please?

**Review by stephybun — COMMENTED** [review]
Apologies @KiPIDesTAN it appears I was overzealous when doing some housekeeping in our SDK and removed the `2021-07-01-preview` version since it wasn't in use in the provider at the time. 

Leaving this comment here so maintainers know not to merge until https://github.com/hashicorp/pandora/pull/4134 has been merged and reincorporated in the provider.

**Comment by KiPIDesTAN** [comment]
@katbyte , any chance you can look at this again? The pandora item mentioned above has been merged back into main.

**Comment by KiPIDesTAN** [comment]
@stephybun , thanks. Just finished the rebase and should be good.

File: internal/services/monitor/monitor_private_link_scope_resource.go [hunk 2]
```diff
@@ -47,6 +48,30 @@ func resourceMonitorPrivateLinkScope() *pluginsdk.Resource {
...
+			"ingestion_access_mode": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  string(privatelinkscopesapis.AccessModeOpen),
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						string(privatelinkscopesapis.AccessModeOpen),
+						string(privatelinkscopesapis.AccessModePrivateOnly),
+					},
+					false),
```
Comment (stephybun) [inline internal/services/monitor/monitor_private_link_scope_resource.go:60]: We can replace this with 
```suggestion
				ValidateFunc: validation.StringInSlice(privatelinkscopesapis.PossibleValuesForAccessMode(), false),
```

File: internal/services/monitor/monitor_private_link_scope_resource.go [hunk 3]
```diff
@@ -47,6 +48,30 @@ func resourceMonitorPrivateLinkScope() *pluginsdk.Resource {
...
+			"query_access_mode": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  string(privatelinkscopesapis.AccessModeOpen),
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						string(privatelinkscopesapis.AccessModeOpen),
+						string(privatelinkscopesapis.AccessModePrivateOnly),
+					},
+					false),
```
Comment (stephybun) [inline internal/services/monitor/monitor_private_link_scope_resource.go:72]: Same here 
```suggestion
				ValidateFunc: validation.StringInSlice(privatelinkscopesapis.PossibleValuesForAccessMode(), false),
```

File: internal/services/monitor/monitor_private_link_scope_resource.go [hunk 4]
```diff
@@ -47,6 +48,30 @@ func resourceMonitorPrivateLinkScope() *pluginsdk.Resource {
 				ValidateFunc: validate.PrivateLinkScopeName,
 			},
 
+			"ingestion_access_mode": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				Default:  string(privatelinkscopesapis.AccessModeOpen),
```
Comment (stephybun) [inline internal/services/monitor/monitor_private_link_scope_resource.go:54]: Could you confirm whether is this the default value returned by the API?

File: internal/services/monitor/monitor_private_link_scope_resource_test.go [hunk 5]
```diff
@@ -134,21 +135,27 @@ func (r MonitorPrivateLinkScopeResource) requiresImport(data acceptance.TestData
 resource "azurerm_monitor_private_link_scope" "import" {
   name                = azurerm_monitor_private_link_scope.test.name
   resource_group_name = azurerm_monitor_private_link_scope.test.resource_group_name
+
+  ingestion_access_mode = "Open"
+  query_access_mode     = "Open"
```
Comment (stephybun) [inline internal/services/monitor/monitor_private_link_scope_resource_test.go:140]: These aren't needed here
```suggestion
```

File: website/docs/r/monitor_private_link_scope.html.markdown [hunk 6]
```diff
@@ -32,6 +35,10 @@ The following arguments are supported:
 
 * `resource_group_name` - (Required) The name of the Resource Group where the Azure Monitor Private Link Scope should exist. Changing this forces a new resource to be created.
 
+* `ingestion_access_mode` - (Optional) The default ingestion access mode for the associated private endpoints in scope. Possible values `Open` and `PrivateOnly`. Omitting this field sets this value to `Open`.
+
+* `query_access_mode` - (Optional) The default query access mode for hte associated private endpoints in scope. Possible values `Open` and `PrivateOnly`. Omitting this field sets this value to `Open`.
```
Comment (stephybun) [inline website/docs/r/monitor_private_link_scope.html.markdown:40]:
```suggestion
* `ingestion_access_mode` - (Optional) The default ingestion access mode for the associated private endpoints in scope. Possible values are `Open` and `PrivateOnly`. Defaults to `Open`.

* `query_access_mode` - (Optional) The default query access mode for hte associated private endpoints in scope. Possible values are `Open` and `PrivateOnly`. Defaults to `Open`.
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @KiPIDesTAN. Could you take a look at the suggestions and questions left in-line? Once that's done this should be good to go.

File: internal/services/monitor/monitor_private_link_scope_resource.go (hunk 4 above)
Comment (KiPIDesTAN) [reply internal/services/monitor/monitor_private_link_scope_resource.go:54]: That's correct. Open ingestion mode is the default.

**Review by stephybun — APPROVED** [review]
Thanks @KiPIDesTAN. Tests look good LGTM 🚀

**Pull request merged**